src/commands/mergecellscommand.js

```diff
@@ -199,16 +199,45 @@ function getBiggestRectangleArea( rows, columns ) {
 	return ( lastRow - firstRow + 1 ) * ( lastColumn - firstColumn + 1 );
 }
 
+// Checks if selection does not mix header (column or row) with other cells.
+//
+// In the below table a valid selection is such that consist cells with the same letter.
+// So, a-a (same heading row and column) or d-d (body cells) are valid while c-d or a-b are not.
+//
+//    header columns
+//     ↓   ↓
+//   ┌───┬───┬───┬───┐
+//   │ a │ a │ b │ b │  ← header row
+//   ├───┼───┼───┼───│
+//   │ c │ c │ d │ d │
+//   ├───┼───┼───┼───│
+//   │ c │ c │ d │ d │
+//   └───┴───┴───┴───┘
+//
 function areCellInTheSameTableSection( tableCells ) {
 	const table = findAncestor( 'table', tableCells[ 0 ] );
 
 	const rowIndexes = getRowIndexes( tableCells );
 	const headingRows = parseInt( table.getAttribute( 'headingRows' ) || 0 );
 
-	const firstCellIsInBody = rowIndexes.first > headingRows - 1;
-	const lastCellIsInBody = rowIndexes.last > headingRows - 1;
+	// Calculating row indexes is a bit cheaper so if this check fails we can't merge.
+	if ( !areIndexesInSameSection( rowIndexes, headingRows ) ) {
+		return false;
+	}
+
+	const headingColumns = parseInt( table.getAttribute( 'headingColumns' ) || 0 );
+	const columnIndexes = getColumnIndexes( tableCells );
```

Comment:
I know that this is not the scope of this PR but `getColumnIndexes` confused me a little, maybe it should be sth like `getFirstLastColumnIndexes`? Same for `getRowIndexes` -> `getFirstLastRowIndexes`.

Reply:
> I know that this is not the scope of this PR

Yep, for me, it is outside a PR's scope. I'm fine with opening a new PR/issue for this. But here I didn't touch that code.